### PR TITLE
Added Sub method

### PR DESCRIPTION
**Pull request by ctessum**

### CODE REVIEW

File: badunit/badunit.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
+package badunit
+
+import "github.com/ctessum/unit"
```
Comment (sbinet) [inline badunit/badunit.go]: this doesn't look right :)

File: dimensions.go [hunk 2]
```diff
@@ -1,7 +1,3 @@
-// Copyright ©2013 The gonum Authors. All rights reserved.
```
Comment (sbinet) [inline dimensions.go]: why is this being removed?

File: badunit/badunit.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
+package badunit
```
Comment (sbinet) [inline badunit/badunit.go]: License header

File: unit.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
+// Copyright ©2013 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package unit
+
+import "fmt"
+
+// Unit is a type a value with generic SI units. Most useful for
```
Comment (sbinet) [inline unit.go]: `s/Unit is a type a value/Unit describes a value in generic SI units/`
or something along these lines?
(not a native english speaker, but I had issues parsing that sentence)

File: unit.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+	formatted  string
+	value      float64
+}
+
+// New creates a new variable of type Unit which has the value
+// specified by value and the dimensions specified by the
+// base units struct. The value is always in SI Units.
+//
+// Example: To create an acceleration of 3 m/s^2, one could do
+// myvar := CreateUnit(3.0, &Dimensions{unit.LengthDim: 1, unit.TimeDim: -2})
```
Comment (sbinet) [inline unit.go]: `s/CreateUnit/New/`

File: unit.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+func New(value float64, d Dimensions) *Unit {
```
Comment (sbinet) [inline unit.go]: I surmise this is a leftover from the `Uniter` interface but: why the `*Unit` indirection?
wouldn't returning it by value put less pressure on the GC?

File: unit.go [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+	u := &Unit{
+		dimensions: make(map[Dimension]int),
```
Comment (sbinet) [inline unit.go]: `s/make(map[Dimension]int)/make(Dimensions)/`

File: unit.go [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+			u.dimensions[key] = val
+		}
+	}
+	return u
+}
+
+// Clone makes a copy of a unit.
+func (u *Unit) Clone() *Unit {
+	o := &Unit{
+		dimensions: make(map[Dimension]int),
```
Comment (sbinet) [inline unit.go]: `s/make(map[Dimension]int)/make(Dimensions)/`

File: unit.go [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+		if val != 0 {
+			o.dimensions[key] = val
+		}
+	}
+	return o
+}
+
+// DimensionsMatch checks if the dimensions of two *Units are the same.
+func DimensionsMatch(a, b *Unit) bool {
+	aUnit := a
```
Comment (sbinet) [inline unit.go]: I don't see the need to introduce `aUnit` nor `bUnit`.

File: unit.go [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+	return true
+}
+
+// Matches checks if the two sets of dimensions are the same.
+func (d Dimensions) Matches(d2 Dimensions) bool {
+	if len(d) != len(d2) {
+		return false
+	}
+	for key, val := range d {
+		if d2[key] != val {
```
Comment (sbinet) [inline unit.go]: what if `d2` does not have key `key` but another one that `d` hasn't ?
shouldn't we also use `v2,ok := d2[key]` and/or also range-loop over `d2` ?

File: unit.go [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+}
+
+// Sub subtracts the second-through-last function arguments
+// from the first function argument.
+// It panics if the units of the arguments don't match.
+// Nil arguments cause a panic.
+func Sub(u ...*Unit) *Unit {
+	return operatePanic(func(o, uu *Unit, i int) {
+		if !DimensionsMatch(o, uu) {
+			panic(fmt.Errorf("Mismatched dimensions in subtraction: "+
```
Comment (sbinet) [inline unit.go]: prefix with `unit: `

File: unit.go [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+				o.Dimensions(), i, uu.Dimensions()))
+		}
+		o.value -= uu.value
+	}, u)
+}
+
+// Sub subtracts uu from u, modifying u instead of creating a copy.
+func (u *Unit) Sub(uu *Unit) {
+	if !DimensionsMatch(u, uu) {
+		panic(fmt.Errorf("Mismatched dimensions in subtraction: "+
```
Comment (sbinet) [inline unit.go]: prefix with `unit: `

File: unit.go [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+	u.formatted = ""
+}
+
+// Max returns the maximum among function arguments.
+// It panics if the units of the arguments don't match.
+// Any nil values are ignored.
+func Max(u ...*Unit) *Unit {
+	return operateIgnore(func(o, uu *Unit, i int) {
+		if !DimensionsMatch(o, uu) {
+			panic(fmt.Errorf("Mismatched dimensions in addition: "+
```
Comment (sbinet) [inline unit.go]: prefix with `unit: `

File: unit.go [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+	}, u)
+}
+
+// Min returns the minimum among function arguments.
+// It panics if the units of the arguments don't match.
+// Any nil values are ignored.
+func Min(u ...*Unit) *Unit {
+	return operateIgnore(func(o, uu *Unit, i int) {
+		if !DimensionsMatch(o, uu) {
+			panic(fmt.Errorf("Mismatched dimensions in addition: "+
```
Comment (sbinet) [inline unit.go]: prefix with `unit: `

**Comment by kortschak** [comment]
What's the rationale for this PR?

**Comment by ctessum** [comment]
I had forgotten about this. I made this pull request over a year ago to just add a "Sub" method for doing subtraction. The rationale for that should be obvious enough.

Since then I have been using this package for a project, and while using it I became frustrated with the combination of statically and dynamically specified units types, and started to feel that because the statically specified types are not able to do multiplication or division, they are pretty much worthless. Static typing is obviously desirable in theory but this implementation wasn't working for me at all and I haven't been able to come up with an elegant solution that involves static typing. The only thing I've thought of is to add a compiler tag to turn on or off dynamic unit checking, but that doesn't seem ideal.

So in my fork I deleted the static types, expanded the dynamic type, and added a "badunit" subpackage for converting from other unit systems into SI. 

Meanwhile, I had forgotten that this PR was still tracking the master branch of my fork. Perhaps the best way forward would be to close this PR (I will try to implement the suggestions regardless), and I can open new PR(s) with the changes in my branch that are deemed worthy of integrating into the main repository. Let me know if any of the changes are of interest for inclusion, and if so which ones. I can start a discussion on the google group if that seems more appropriate.

**Comment by kortschak** [comment]
Yes, close this and reopen another with the Sub addition.

**Pull request closed without merging**